src/utils/domains.ts

```diff
@@ -0,0 +1,162 @@
+import { placeholder } from './placeholder';
+import { Domain, Datapoint, DomainPriority } from '../external';
+import { ScaledSeries, DataDomains } from '../internal';
+
+export const isEqual = (a?: Domain, b?: Domain): boolean => {
+  if (a === b) {
+    return true;
+  }
+  if (!!a !== !!b) {
+    return false;
+  }
+  // This should never happen, but TypeScript isn't smart enough to see what
+  // happened with the check above.
+  if (!a || !b) {
+    return false;
+  }
+  return a[0] === b[0] && a[1] === b[1];
+};
+
+export const newDomain = (
+  start: number,
+  end: number,
+  priority: DomainPriority
+): Domain => {
+  const domain: [number, number] & { priority?: DomainPriority } = [start, end];
+  domain.priority = priority;
+  domain.toString = () => `[${start}, ${end}] (priority: ${priority})`;
+  return domain as Domain;
+};
+
+export const copyDomain = (
+  domain: Domain,
+  priority?: DomainPriority
+): Domain => {
+  return newDomain(domain[0], domain[1], priority || domain.priority);
+};
+
+export const copyDataDomains = ({ time, x, y }: DataDomains): DataDomains => ({
+  time: copyDomain(time),
+  x: copyDomain(x),
+  y: copyDomain(y),
+});
+
+export const highestPriorityDomain = (
+  ...domains: Array<Domain | undefined>
+): Domain | undefined => {
+  const filtered: Domain[] = domains.filter(
+    domain => domain !== undefined
+  ) as Domain[];
+  if (filtered.length === 0) {
+    return undefined;
+  }
+  if (filtered.length === 1) {
+    return filtered[0];
+  }
+
+  // Note that sort() updates the array (but also returns it) so filtered is
+  // also sorted by this. However, the returned variable is renamed in the
+  // interest of readability.
+  const sorted = filtered.sort((a: Domain, b: Domain) => {
+    return b.priority - a.priority;
+  });
+  return sorted[0];
+};
+
+const withPadding = (extent: [number, number]): Domain => {
+  const diff = extent[1] - extent[0];
+  if (Math.abs(diff) < 1e-3) {
+    if (extent[0] === 0) {
+      // If 0 is the only value present in the series, hard code domain.
+      return newDomain(-0.25, 0.25, DomainPriority.CALCULATED);
+    }
+    const domain = [(1 / 2) * extent[0], (3 / 2) * extent[0]];
+    if (domain[1] < domain[0]) {
+      return newDomain(domain[1], domain[0], DomainPriority.CALCULATED);
+    }
+    // @ts-ignore - this is valid -- why are you complaining?
+    return domain;
+  }
+  return newDomain(
+    extent[0] - diff * 0.025,
```

Comment:
Just wondering what is the reasoning behind `0.025`, `0.25`, `1e-3`?

Reply:
This was just a buffer that "seemed right". I don't know the reasoning behind it.